### PR TITLE
Symex: resolve pointer comparisons using the value-set

### PR DESCRIPTION
**Pull request by owen-mc-diffblue**
This allows us to evaluate boolean expressions (including GOTO conditions) that test pointer equality or inequality (particularly "ptr == null" or "!= null") where the value-set contains sufficient information. We can already handle simple cases where `ptr` has a single known alias via constant propagation and aliasing, but not cases where `ptr` is non-constant, for example because it is drawn from an array.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-symex/goto_symex.h
Comment (owen-mc-diffblue) [inline src/goto-symex/goto_symex.h]: Any suggestions for a better name?

File: src/goto-symex/goto_symex.h
Comment (owen-mc-diffblue) [inline src/goto-symex/goto_symex.h]: This function might fit better somewhere else - perhaps in `value_set_dereferencet`, since it's mainly a wrapper around `value_set_dereferencet::build_reference_to`?

File: src/goto-symex/symex_goto.cpp
Comment (owen-mc-diffblue) [inline src/goto-symex/symex_goto.cpp]: anything else I should be recursing over, other than AND, OR and NOT?

File: src/goto-symex/symex_goto.cpp
Comment (owen-mc-diffblue) [inline src/goto-symex/symex_goto.cpp]: Currently it assumes the symbol is on the lhs and the constant is on the rhs. I will fix that up.

File: src/goto-symex/symex_main.cpp
Comment (owen-mc-diffblue) [inline src/goto-symex/symex_main.cpp]: This code was in `goto_symext::try_filter_value_sets` and I've pulled it out into a function to reuse it. Perhaps I'll make that a separate commit.

File: regression/cbmc/symex_should_evaluate_simple_pointer_conditions/test.c
Comment (smowton) [inline regression/cbmc/symex_should_evaluate_simple_pointer_conditions/test.c]: Bear in mind these constant comparisons with `c` may well be resolved very early as `c` is only assigned once and never address-taken.

File: src/goto-symex/symex_goto.cpp
Comment (smowton) [inline src/goto-symex/symex_goto.cpp]: Handle the reverse case too (constant OP symbol)

File: src/goto-symex/symex_goto.cpp
Comment (smowton) [inline src/goto-symex/symex_goto.cpp]: Assert that `symbol_expr` is already L1 at least

File: src/goto-symex/symex_goto.cpp
Comment (smowton) [inline src/goto-symex/symex_goto.cpp]: Surprising! Null is usually represented as an `ID_constant` with value = NULL -- suspect you need to add tests for this

File: src/goto-symex/symex_goto.cpp
Comment (smowton) [inline src/goto-symex/symex_goto.cpp]: Give some brief comments setting out the logic here, as `!constant_found` and `value_set_elements.size() == 1` don't immediately chime with "x in { a, b, c }" and "x in { x }" respectively.

**Review by smowton — COMMENTED** [review]
Looks good except I believe the null case is broken, and at least needs tests.

File: src/goto-symex/symex_goto.cpp
Comment (smowton) [reply src/goto-symex/symex_goto.cpp]: (and test this)

File: src/goto-symex/goto_symex.h
Comment (smowton) [reply src/goto-symex/goto_symex.h]: I like this one!

File: src/goto-symex/goto_symex.h
Comment (smowton) [reply src/goto-symex/goto_symex.h]: Proooobably. No strong opinion.

File: src/goto-symex/symex_goto.cpp
Comment (smowton) [reply src/goto-symex/symex_goto.cpp]: No I think that's enough

Comment (smowton) [reply src/goto-symex/symex_goto.cpp]: Oh wait, how about the arguments of a boolean-typed `if_exprt`

Comment (smowton) [reply src/goto-symex/symex_goto.cpp]: (e.g. `x == y ? z == w : a == b`)

Comment (smowton) [reply src/goto-symex/symex_goto.cpp]: Or indeed the first arg of an `if_exprt` in any context. Total recursion looking for `==` and `!=` nodes might be the simplest thing.

Comment (smowton) [reply src/goto-symex/symex_goto.cpp]: I 100% think these comments through before beginning to write FYI

File: regression/cbmc/symex_should_evaluate_simple_pointer_conditions/test.c
Comment (owen-mc-diffblue) [reply regression/cbmc/symex_should_evaluate_simple_pointer_conditions/test.c]: I've changed them to other pointer comparisons. And I now take the address of all the integers at the beginning, if that's a thing that symex cares about.

**Comment by owen-mc-diffblue** [comment]
I've addressed all of @smowton's review comments on the draft PR and got CI working (touch wood).

File: regression/cbmc/symex_should_evaluate_simple_pointer_conditions/test.c
Comment (smowton) [inline regression/cbmc/symex_should_evaluate_simple_pointer_conditions/test.c]: correct -> matching and incorrect -> non-matching everywhere

File: regression/cbmc/symex_should_evaluate_simple_pointer_conditions/test.c
Comment (smowton) [inline regression/cbmc/symex_should_evaluate_simple_pointer_conditions/test.c]: the -> a

File: regression/cbmc/symex_should_evaluate_simple_pointer_conditions/test.desc
Comment (smowton) [inline regression/cbmc/symex_should_evaluate_simple_pointer_conditions/test.desc:56]: Suggest adding conditions that check you check for everything you should, by making sure `unreachable_1[3-9])` and `unreachable[2-9][0-9]` for example don't occur anywhere (i.e., if someone added a test but not a cross-check this regex would complain)

File: src/goto-symex/symex_goto.cpp
Comment (smowton) [inline src/goto-symex/symex_goto.cpp]: If you're not sure this is simplified you might get redundant stacked typecast ops

File: src/goto-symex/symex_goto.cpp
Comment (smowton) [inline src/goto-symex/symex_goto.cpp]: No need for braces with one-liner `if`s

File: src/goto-symex/symex_goto.cpp
Comment (smowton) [inline src/goto-symex/symex_goto.cpp]: Instead of calling this twice, suggest something like https://github.com/diffblue/cbmc/blob/develop/src/goto-symex/goto_state.cpp#L67

File: src/goto-symex/symex_goto.cpp
Comment (smowton) [inline src/goto-symex/symex_goto.cpp]: How about `try_evaluate_pointer_comparison`, singular?

File: src/goto-symex/symex_goto.cpp
Comment (smowton) [inline src/goto-symex/symex_goto.cpp]: I don't think there's much gained by the two functions (with_check vs. without) since they're both only called in one place

**Comment by owen-mc-diffblue** [comment]
Note also that the last two commits break sharing everywhere, so we'll have to rethink how to do that.

**Comment by smowton** [comment]
Suggested change: https://github.com/smowton/cbmc/commit/1daa2c60056213d2eddce6f9e3a2945a139d81b2

That removes `renamed_vectort` and instead uses `depth_begin/depth_end`, which avoid sharing breaks when no changes are made, to selectively alter the renamed expression, only permitting changes where the renaming level is consistent. @romainbrenguier are you happy with this?

@owen-jones-diffblue note that also applies a couple of my suggestions above since it made it easier to code :)

File: src/goto-symex/symex_goto.cpp
Comment (owen-mc-diffblue) [reply src/goto-symex/symex_goto.cpp]: I prefer to use them in all cases

**Comment by romainbrenguier** [comment]
@smowton 
> Suggested change: smowton@1daa2c6
> That removes renamed_vectort and instead uses depth_begin/depth_end, which avoid sharing breaks when no changes are made, to selectively alter the renamed expression, only permitting changes where the renaming level is consistent. @romainbrenguier are you happy with this?

Yes that looks better. What I'm not sure about is the fact that the mutator function takes an `exprt` as argument rather that a `renamedt`. That way you avoid calling constructors on all intermediary expressions, but I don't think it's nice for the interface. Maybe we could make `renamedt` privately inherit from exprt so that we can use `static_cast` instead of a constructor, so that conversion would be cheap and the interface nicer?
Here is the change I propose (if you want to use it): https://github.com/diffblue/cbmc/pull/4454

**Comment by owen-mc-diffblue** [comment]
Thanks @smowton, that's much better. I've cherry-picked it and squashed it in, along with fixes for all the things you mentioned. I did add an extra commit which makes `selectively_mutate` a free function (which is a friend of `renamedt`), since it doesn't make sense to be using depth iterators for anything when `underlyingt` is derived from `exprt`. I'm not sure how this interacts with @romainbrenguier's suggestion.

There are three failing tests. The first is `exception_cleanup`, which is failing because symex can now tell that `@inflight_exception` is not null, so the test needs to be changed. I'm not totally sure what it should now be. The failing lines are:
```
1 remaining after simplification$
```
in test.desc and
```
file Test\.java line 6 function java::Test.main:\(I\)V
no uncaught exception
```
in vccs.desc.

The other two are nondet_elements_longer_lists and nondet_elements_longer_lists_global, which are failing in the same way. I'm looking into that at the moment.

**Comment by owen-mc-diffblue** [comment]
For nondet_elements_longer_lists and nondet_elements_longer_lists_global, the test is failing to find these two lines: ```
\[test_function.unwind.\d+\] line \d+ unwinding assertion loop 0: SUCCESS
\[test_function.unwind.\d+\] line \d+ unwinding assertion loop 1: SUCCESS
```
Is there a good reason that we would not generate these unwinding assertions? If symex can determine that they are true then does it just not output them in the formula?

**Comment by smowton** [comment]
@owen-jones-diffblue unsure about the unwinding assertions (yet) but the exception one is entirely desirable -- that 1 VCC remaining was a shortcoming which we have now fixed (symex can tell no exception escaped from the whole program). Just amend the test to check *all* conditions are symex'd away.

**Comment by smowton** [comment]
@romainbrenguier I'm not sure I follow, what would your ideal interface look like?

**Comment by tautschnig** [comment]
@owen-jones-diffblue Presumably your changes just result in more constant propagation, and thus those unwinding assertions are resolved by the simplifier already. #2574 does the same, i.e., removes those lines from the test spec.

**Comment by romainbrenguier** [comment]
@smowton 
> @romainbrenguier I'm not sure I follow, what would your ideal interface look like?

I would think having `mutator_functiont` take `renamedt` arguments would be more useful (though in your case it doesn't really matter). For instance imagine we want to selectively apply `simplify` to operands of a `renamedt`, you could do it with a `renamedt -> optinalt<renamedt>` function. I'm not insisting on it, since it is not necessary here, but could be nice in the future.

**Comment by smowton** [comment]
Ah sure, ok. On the one hand the tag is useful if you want to call a function that demands a `renamedt`, but I certainly like not having to make a dedicated object for every subexpression, so your privately-inherit trick looks like a good solution.

**Comment by owen-mc-diffblue** [comment]
I've marked this as `do not merge` until we have some benchmarking data for it

**Comment by owen-mc-diffblue** [comment]
I've cherry-picked ##4454 and added a commit to use it. Before merging this should be cleaned up.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: eb9b8cecade7195d17aa7de83276962168a60daa).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106349820

File: src/util/ssa_expr.h
Comment (tautschnig) [inline src/util/ssa_expr.h]: If adding those, could we please unify the various checking-functions in the same way this has been implemented in `std_expr.h` for all the classes? That is, use `validate_expr` in the above.

File: src/pointer-analysis/value_set_dereference.h
Comment (tautschnig) [inline src/pointer-analysis/value_set_dereference.h]: Please document.

File: regression/cbmc/symex_should_evaluate_simple_pointer_conditions/test.desc
Comment (tautschnig) [inline regression/cbmc/symex_should_evaluate_simple_pointer_conditions/test.desc]: GitHub suggests adding a final newline.

File: src/goto-symex/symex_goto.cpp
Comment (tautschnig) [inline src/goto-symex/symex_goto.cpp]: Use `skip_typecast`?

File: src/pointer-analysis/value_set_dereference.h
Comment (owen-mc-diffblue) [reply src/pointer-analysis/value_set_dereference.h]: What I want to call it is `build_reference_to`, but that's already taken by the function it's wrapping (which doesn't actually build a reference).

The reason I have a wrapper function is that `build_reference_to` makes it easy to get objects, but not pointers, e.g. it gives you `global`, not `&global`, and for null pointers it uses the special value `nil_exprt`, and doesn't give me the `null_pointer_exprt` that I want. That makes sense because I think it's only used for `dereference_rec`, which only cares about dereferenced pointers.  Now that I've read through that function a few more times, I think a better solution would be to add a field to `value_set_dereferencet::valuet` to put `&global` and `null_pointer_exprt` into. What would you say to that?

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 926531eae1c1eeb6c03c9deab5570c422639484e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106591286

File: src/goto-symex/symex_goto.cpp
Comment (owen-mc-diffblue) [reply src/goto-symex/symex_goto.cpp]: I tried to search for such a utility but didn't hit the right combination of words. Next time I'll try looking in `expr_utils.h`.

**Comment by owen-mc-diffblue** [comment]
@tautschnig I've addressed all but one comment. Let me know what you think about my suggested approach to addressing the last comment.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 44d45d645e96fc5787b21c97e20ca4e5e4f400a9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106623135

File: src/pointer-analysis/value_set_dereference.h
Comment (owen-mc-diffblue) [reply src/pointer-analysis/value_set_dereference.h]: I have done this now

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c0e7ab8c6151a9943d901e145acffc7a581c10a8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106938073

**Comment by smowton** [comment]
@tautschnig initial findings from studying this: the test `jbmc/regression/jbmc/exception-cleanup` gives a good example of where this can help, as even with #3619 and #4430 we get an expression where propagation works, but the simplifier isn't clever enough to spot that a unique value can be determined -- in this case we get `guard ? null : (guard ? &x : null)`, which is clearly always null but requires the simplifier to propagate that `guard => false` into the RHS of the conditional, which it doesn't attempt at present.

Here's another example where propagation alone doesn't work:
```c
#include <assert.h>

int main(int argc, char **argv) {
  int x, y;
  int *p = argc == 1 ? 0 : argc == 2 ? &x : &y;
  if(p != 0) {
    if(p == 0) {
      assert(0);
    }
  }
}
```

The redundant null-tests on `p` can't be resolved because by simplification because the knowledge that the null case has already been filtered out is encoded in the value-set but not in the expression itself. This could be done, for example by transforming `p == (argc == 1 ? 0 : argc == 2 ? &x : &y) && p != 0` into `argc == 2 ? &x : &y`, but we currently don't attempt this, and if we did I expect we would either (a) need to switch on local-definitions in the simplifier, which could get quite expensive, or (b) the augmented simplifier would probably only catch particular cases rather than generally propagating conditions into expressions.

In brief I continue to think that at least for the time being the value-set is a more robust way of tracking aliases than trying to propagate all if-expressions encoding aliases and improve the simplifier sufficiently to filter those nested conditionals and solve tests against them. I think this is particularly a problem as in the limiting case this results in spiralling expression sizes and so the cost of `simplify` and `merge_ireps` rises, as I'm currently trying to fight *against* in #4576.

File: src/goto-symex/renaming_level.h
Comment (tautschnig) [inline src/goto-symex/renaming_level.h:114]: I'm a bit surprised by the amount of friendship in here. Couldn't some of these be made static members instead?

**Review by tautschnig — APPROVED** [review]
Looks ok me, except the remaining clang-format problem and, more importantly, the absence of performance data - could this please be provided? I certainly see that this gives symex additional power in resolving some comparisons, but at what cost does this come?

**Comment by smowton** [comment]
cbmc CORE unit tests times: unchanged

**Comment by smowton** [comment]
Large Webgoat example: 55 seconds to reach a particular checkpoint (down 5 seconds)

**Comment by smowton** [comment]
Romain's Sakai benchmark: One benchmark reduced around 15 seconds (from 145 seconds), the remainder tested showed negligible difference.

**Comment by smowton** [comment]
Therefore I'm pretty confident this has no noticeable cost in context of the other fairly-expensive ops we perform in symex.

File: src/goto-symex/renaming_level.h
Comment (smowton) [reply src/goto-symex/renaming_level.h:114]: The gymnastics here are a little beyond me, so I think I'll leave this as-is and welcome any cleanup PRs by people who're better at this game than me :)

File: src/goto-symex/symex_goto.cpp
Comment (peterschrammel) [inline src/goto-symex/symex_goto.cpp:87]: `const &`

File: src/goto-symex/symex_goto.cpp
Comment (peterschrammel) [inline src/goto-symex/symex_goto.cpp]: `const &`

File: src/goto-symex/symex_goto.cpp
Comment (peterschrammel) [inline src/goto-symex/symex_goto.cpp]: `const &`

File: src/goto-symex/symex_goto.cpp
Comment (peterschrammel) [inline src/goto-symex/symex_goto.cpp]: `const &`

**Review by peterschrammel — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: ff6002bb92522460931c3ffc6c3d517e3ac8b0b1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/111618837

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 213a1b25c0eeb10405005619b45ce71eb82490a1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/111750931

**Comment by tautschnig** [comment]
I'm afraid this actually causes spurious counterexamples being reported. I'll try to build a small regression test.

**Comment by tautschnig** [comment]
Actually it's pretty simple:
```
int main()
{
  void *ptr;
  if(ptr == 0)
  {
    return 0;
  }
  __CPROVER_assert(ptr != 0, "null");
}
```
I believe the problem is that `ptr` has never been assigned, and thus the value set has no information about it.

**Comment by smowton** [comment]
Discussion on Slack summary: will revert and re-submit with a fix

**Pull request merged**
